### DISCUSSION TITLE
Add editing functionality to mapbox-gl layer - don’t render everything in SVG.

### DISCUSSION THREAD

**Discussion by aliya-ryan** [digidem/mapeo-desktop#542] [Ideas]
Depends on having the mapbox layer rendered underneath editing layer in Mapeo - due for April milestone.

**Comment by benstadin** [comment]
I'm interested in this feature. Is this still planned?

**Comment by gmaclennan** [comment]
Hi @benstadin yes this is still on the roadmap. Our current plan is to render the map in mapbox-gl, and then on click, render the feature-to-be-edited using the current iD Editor SVG render layer. The steps involved are:

- [ ] Generate realtime vector tiles from osm-p2p-data: https://github.com/digidem/osm-p2p-vector-tile-server (not currently working)
- [ ] Add a [mapbox-gl-js](https://github.com/mapbox/mapbox-gl-js/) render layer to iD
- [ ] Dynamically generate the mapbox style from presets in order to render icons etc.
- [ ] Add interactivity to mapbox-gl layer and re-write SVG layer to only show currently selected features.
- [ ] Figure out what to do with edited but not saved features? Continue to render in SVG?

Longer term the ideal would be for all editing to be done in WebGL, but that would be a lot more work. It may be that Mapbox starts to switch iD from SVG to WebGL at some point.

mapeo-desktop development has been on hold recently as we have focussed on other project priorities, but we are hoping to bring on a new developer soon to start tackling some of these features.

**Comment by benstadin** [comment]
Hi @gmaclennan good to know. Thanks for your quick reply.

Just some random thoughts as an alternative approach to render vector data and mixing it with D3/SVG edit data. I just realized that MapBox is somewhat limited when it comes to filtering data (style filters cannot be implemented as custom functions). 

- Generate vector tiles statically instead of doing it in realtime. Like convert from OSM to GeoJSON, and from GeoJson to MapBox mvt tiles via https://github.com/NYCPlanning/geojson2mvt
- Query the feature on the MapBox map which is currently being edited. Somewhat related code example: https://www.mapbox.com/mapbox-gl-js/example/query-similar-features/
- Alter the queried feature, set key "editing" to in the feature's properties. 
- Add a filter to hide features with editing == true: https://www.mapbox.com/mapbox-gl-js/example/filter-markers/
- Add another filter for key "was-edited"
- Save all edited features in a GeoJson file. And add this file as second data source to MapBox.
- Loop through currently visible features in the main data source and add a tag to those features which are also in the edited features GeoJson. Do this on startup and when moving the map: https://www.mapbox.com/mapbox-gl-js/example/filter-features-within-map-view/

**Comment by gmaclennan** [comment]
@benstadin thanks for this, these are really helpful suggestions.
